### PR TITLE
[1837] add background jobs for publish

### PR DESCRIPTION
**Pull request by misaka**
### Context

We need to add a background worker to process jobs to sync courses to search-and-compare.

### Changes proposed in this pull request

Add `delayed_job`, configure it for the different environments, and modify Docker to build it so that it can be deployed.

### Guidance to review

This PR is stacked ontop of #643 which was not being merged until we could demonstrate that this whole thing works. That PR should be merged into master once this one is approved, and then this one rebased.

### Checklist

- [ ] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: app/services/search_and_compare_api_service.rb [hunk 1]
```diff
@@ -13,32 +13,32 @@ def api
...
-          serialize(course)
-        )
+    attr_reader :response
 
-        response.success?
-      end
+    # NOTE:
+    #   HTTP PUT "/api/courses" accepts a list of courses
+    #   It will only add or update, no DELETION.
+    def sync(course)
```
Comment (defong) [inline app/services/search_and_compare_api_service.rb]: can you change that while you here to reflect its a list

File: bg-find-sync/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+ARG  DOCKER_REPOSITORY
+ARG  MANAGE_COURSES_BACKEND_IMAGE
+ARG  CODE_VERSION=latest
+
+FROM ${DOCKER_REPOSITORY}/${MANAGE_COURSES_BACKEND_IMAGE}:${CODE_VERSION}
+
+CMD bundle exec rails db:migrate && bundle exec bin/delayed_job --queue=find_sync run
```
Comment (defong) [inline bg-find-sync/Dockerfile]: should this do migrate? and not the actual site....

File: docker-compose.yml [hunk 3]
```diff
@@ -11,8 +11,8 @@ services:
     build:
       context: .
       cache_from:
-        - batdevcontainerregistry.azurecr.io/manage-courses-backend:latest
-    image: batdevcontainerregistry.azurecr.io/manage-courses-backend:latest
+        - dfedigital/manage-courses-backend:latest
+    image: dfedigital/manage-courses-backend:latest
```
Comment (tvararu) [inline docker-compose.yml]: This seems unrelated to the other changes, changing the container registry for the mcbe image.

Comment (misaka) [reply docker-compose.yml]: I just didn't like the ider of them being on separate registries. May just revert for now.

**Comment by misaka** [comment]
Just to comment that we won't be able to merge this, actually, until we migrate to the new prod environment, since we won't be building this image in Travis.

File: bg-find-sync/Dockerfile (hunk 2 above)
Comment (misaka) [reply bg-find-sync/Dockerfile]: > should this do migrate? and not the actual site....

I think it's ok to do the migrate here, but we'll have to see if there's any race conditions. I'd rather not require the web app to be started before this container can be started.

File: app/services/search_and_compare_api_service.rb (hunk 1 above)
Comment (misaka) [reply app/services/search_and_compare_api_service.rb]: Yup, will change.

**Comment by misaka** [comment]
> 374
> pull access denied for dfedigital/manage-courses-backend, repository does not exist or may require 'docker login'

We're not going to be able to deploy this with Travis anyway, so that error is moot. It builds on Azure DevOps, although other tests fail.

File: spec/jobs/sync_courses_to_find_job_spec.rb [hunk 4]
```diff
@@ -15,9 +16,9 @@
   end
 
   it 'syncs using the SearchAndCompareAPIService' do
-    described_class.perform_now(course)
+    expect_any_instance_of(SearchAndCompareAPIService::Request)
+      .to receive(:sync).with([course])
 
-    expect(SearchAndCompareAPIService::Request)
-      .to have_received(:sync).with([course])
+    described_class.perform_now(course)
```
Comment (defong) [inline spec/jobs/sync_courses_to_find_job_spec.rb]: Anything interesting happens when there is a http exception?

**Comment by timabell** [comment]
[Add background job process](https://trello.com/c/AeyAxW2B/1837-add-background-job-process)

**Comment by timabell** [comment]
> We're not going to be able to deploy this with Travis anyway

@misaka can we close this PR then until we're ready?

**Comment by misaka** [comment]
I'm on the fence about that. The implementation isn't going to change, so it's not like it'll need a re-review. Will people be happy to re-review?

File: bg-find-sync/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,7 @@
+ARG  DOCKER_REPOSITORY
+ARG  MANAGE_COURSES_BACKEND_IMAGE
+ARG  CODE_VERSION=latest
+
+FROM ${DOCKER_REPOSITORY}/${MANAGE_COURSES_BACKEND_IMAGE}:${CODE_VERSION}
```
Comment (defong) [inline bg-find-sync/Dockerfile]: can this not inherit from `batdevcontainerregistry.azurecr.io/manage-courses-backend:latest`

it looks likes to me that we are building a image

**Comment by defong** [comment]
git rebase master, hence forced push

Need to look at old man @misaka code

File: bg-find-sync/Dockerfile (hunk 2 above)
Comment (misaka) [inline bg-find-sync/Dockerfile]: This duplicates `Dockerfile.bg-find-sync`. We should only have one `Dockerfile` for `bg-find-sync`.

Comment (rizzkhan1) [reply bg-find-sync/Dockerfile]: removed Dockerfile.bg-find-sync
